In [1]:
# ETAPA 2 - COMPREENSÃO E DIAGNÓSTICO DOS DADOS

# Importação das bibliotecas principais
import pandas as pd
import numpy as np

# Configurações para melhorar a visualização no Jupyter
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.float_format", "{:,.2f}".format)

# Caminho do arquivo CSV
caminho_arquivo = "dados-empresa.csv"

# Carregamento da base de dados
df = pd.read_csv(caminho_arquivo)

# Verificar quantidade de linhas e colunas
linhas, colunas = df.shape

print(f"A base possui {linhas} linhas e {colunas} colunas.")

# Listar todas as colunas da base
df.columns.tolist()

A base possui 1020 linhas e 27 colunas.


['id_venda',
 'data',
 'mes',
 'trimestre',
 'cliente',
 'tipo_cliente',
 'tempo_relacionamento_meses',
 'faixa_renda',
 'segmento',
 'produto',
 'categoria',
 'marca',
 'linha_produto',
 'quantidade',
 'preco_unitario',
 'desconto',
 'receita',
 'custo_unitario',
 'lucro',
 'regiao',
 'cidade',
 'canal_venda',
 'vendedor',
 'meta_venda',
 'atingiu_meta',
 'prazo_entrega_dias',
 'status_entrega']

In [3]:
# Verificar os tipos de dados identificados automaticamente pelo Pandas
df.dtypes

id_venda                        int64
data                              str
mes                             int64
trimestre                       int64
cliente                           str
tipo_cliente                      str
tempo_relacionamento_meses      int64
faixa_renda                       str
segmento                          str
produto                           str
categoria                         str
marca                             str
linha_produto                     str
quantidade                      int64
preco_unitario                float64
desconto                      float64
receita                       float64
custo_unitario                float64
lucro                         float64
regiao                            str
cidade                            str
canal_venda                       str
vendedor                          str
meta_venda                    float64
atingiu_meta                      str
prazo_entrega_dias              int64
status_entre

In [4]:
# Criar uma tabela resumida com informações sobre cada coluna
diagnostico_colunas = pd.DataFrame({
    "coluna": df.columns,
    "tipo_pandas": df.dtypes.values,
    "valores_nao_nulos": df.notna().sum().values,
    "valores_nulos": df.isna().sum().values,
    "percentual_nulos": (df.isna().mean().values * 100).round(2),
    "valores_unicos": df.nunique(dropna=True).values
})

diagnostico_colunas

,coluna,tipo_pandas,valores_nao_nulos,valores_nulos,percentual_nulos,valores_unicos
0,id_venda,int64,1020,0,0.00,1000
1,data,str,1020,0,0.00,338
2,mes,int64,1020,0,0.00,12
3,trimestre,int64,1020,0,0.00,4
4,cliente,str,1020,0,0.00,199
5,tipo_cliente,str,1020,0,0.00,2
6,tempo_relacionamento_meses,int64,1020,0,0.00,60
7,faixa_renda,str,969,51,5.00,3
8,segmento,str,1020,0,0.00,2
9,produto,str,1020,0,0.00,8


In [2]:
# Criar uma tabela resumida com informações sobre cada coluna
diagnostico_colunas = pd.DataFrame({
    "coluna": df.columns,
    "tipo_pandas": df.dtypes.values,
    "valores_nao_nulos": df.notna().sum().values,
    "valores_nulos": df.isna().sum().values,
    "percentual_nulos": (df.isna().mean().values * 100).round(2),
    "valores_unicos": df.nunique(dropna=True).values
})

diagnostico_colunas

,coluna,tipo_pandas,valores_nao_nulos,valores_nulos,percentual_nulos,valores_unicos
0,id_venda,int64,1020,0,0.00,1000
1,data,str,1020,0,0.00,338
2,mes,int64,1020,0,0.00,12
3,trimestre,int64,1020,0,0.00,4
4,cliente,str,1020,0,0.00,199
5,tipo_cliente,str,1020,0,0.00,2
6,tempo_relacionamento_meses,int64,1020,0,0.00,60
7,faixa_renda,str,969,51,5.00,3
8,segmento,str,1020,0,0.00,2
9,produto,str,1020,0,0.00,8


In [3]:
# Colunas com valores faltantes
colunas_com_nulos = ["faixa_renda", "preco_unitario", "desconto", "custo_unitario"]

# Verificar quantas linhas possuem pelo menos um valor nulo nessas colunas
linhas_com_algum_nulo = df[colunas_com_nulos].isna().any(axis=1).sum()

# Verificar quantas linhas possuem todos esses campos nulos ao mesmo tempo
linhas_com_todos_nulos = df[colunas_com_nulos].isna().all(axis=1).sum()

print(f"Linhas com pelo menos um nulo nessas colunas: {linhas_com_algum_nulo}")
print(f"Linhas com todos esses campos nulos ao mesmo tempo: {linhas_com_todos_nulos}")

Linhas com pelo menos um nulo nessas colunas: 51
Linhas com todos esses campos nulos ao mesmo tempo: 51


In [4]:
# Verificar duplicidades completas
qtd_duplicadas = df.duplicated().sum()

print(f"Quantidade de linhas completamente duplicadas: {qtd_duplicadas}")

Quantidade de linhas completamente duplicadas: 20


In [8]:
# Verificar duplicidade no identificador da venda
qtd_ids_duplicados = df["id_venda"].duplicated().sum()
qtd_ids_unicos = df["id_venda"].nunique()

print(f"Quantidade de IDs duplicados: {qtd_ids_duplicados}")
print(f"Quantidade de IDs únicos: {qtd_ids_unicos}")

Quantidade de IDs duplicados: 20
Quantidade de IDs únicos: 1000


In [ ]:
# Exibir exemplos de linhas duplicadas
duplicadas = df[df.duplicated(keep=False)].sort_values("id_venda")

duplicadas.head(20)

In [11]:
# Converter a coluna data para datetime apenas para diagnóstico
# Nesta etapa ainda não vamos alterar definitivamente a base
datas_convertidas = pd.to_datetime(df["data"], errors="coerce")

# Verificar período da base
data_minima = datas_convertidas.min()
data_maxima = datas_convertidas.max()

print(f"Data mínima: {data_minima.date()}")
print(f"Data máxima: {data_maxima.date()}")
print(f"Datas inválidas: {datas_convertidas.isna().sum()}")

Data mínima: 2024-01-01
Data máxima: 2024-12-30
Datas inválidas: 0


In [12]:
# Verificar se a coluna mes corresponde ao mês extraído da data
inconsistencia_mes = (datas_convertidas.dt.month != df["mes"]).sum()

# Verificar se a coluna trimestre corresponde ao trimestre extraído da data
inconsistencia_trimestre = (datas_convertidas.dt.quarter != df["trimestre"]).sum()

print(f"Inconsistências na coluna mês: {inconsistencia_mes}")
print(f"Inconsistências na coluna trimestre: {inconsistencia_trimestre}")

Inconsistências na coluna mês: 0
Inconsistências na coluna trimestre: 0


In [ ]:
# Selecionar colunas categóricas
colunas_categoricas = df.select_dtypes(include="str").columns

# Exibir quantidade de valores únicos por coluna categórica
for coluna in colunas_categoricas:
    print(f"\nColuna: {coluna}")
    print(f"Quantidade de valores únicos: {df[coluna].nunique(dropna=True)}")
    print(df[coluna].value_counts(dropna=False))

In [15]:
# Criar tabela cruzando cidade e região
cidade_regiao = pd.crosstab(df["cidade"], df["regiao"])

cidade_regiao

regiao,Centro-Oeste,Exterior,Nordeste,Norte,Sudeste,Sul
cidade,,,,,,
Belém,0,5,0,89,0,0
Brasília,94,2,0,0,0,0
Curitiba,0,7,0,0,0,105
Goiânia,105,2,0,0,0,0
Manaus,0,2,0,118,0,0
Porto Alegre,0,1,0,0,0,102
Recife,0,2,92,0,0,0
Rio de Janeiro,0,3,0,0,99,0
Salvador,0,4,85,0,0,0


In [16]:
# Estatística descritiva das colunas numéricas
df.describe().T

,count,mean,std,min,25%,50%,75%,max
id_venda,"1,020.00",502.94,288.76,1.00,252.75,505.50,754.25,"1,000.00"
mes,"1,020.00",6.47,3.59,1.00,3.00,6.00,10.00,12.00
trimestre,"1,020.00",2.49,1.16,1.00,1.00,2.00,4.00,4.00
tempo_relacionamento_meses,"1,020.00",16.67,18.43,0.00,2.00,5.00,31.00,59.00
quantidade,"1,020.00",6.02,3.16,0.00,3.00,6.00,9.00,11.00
preco_unitario,969.00,"2,917.55","1,728.89",81.45,"1,428.08","2,784.00","4,452.69","5,986.24"
desconto,969.00,0.13,0.07,0.00,0.07,0.13,0.19,0.25
receita,"1,020.00","15,124.91","12,758.54",61.90,"4,615.27","11,167.85","22,705.23","64,052.01"
custo_unitario,969.00,"2,030.47","1,262.81",46.06,988.87,"1,906.49","3,033.53","5,178.20"
lucro,"1,020.00","3,070.05","4,059.18","-5,323.11",421.54,"1,698.83","4,272.87","22,781.83"


In [17]:
# Verificar possíveis valores inválidos nas principais colunas numéricas
valores_invalidos = {
    "quantidade <= 0": (df["quantidade"] <= 0).sum(),
    "preco_unitario <= 0": (df["preco_unitario"] <= 0).sum(),
    "desconto < 0": (df["desconto"] < 0).sum(),
    "desconto > 1": (df["desconto"] > 1).sum(),
    "custo_unitario <= 0": (df["custo_unitario"] <= 0).sum(),
    "receita < 0": (df["receita"] < 0).sum(),
    "lucro < 0": (df["lucro"] < 0).sum(),
    "meta_venda <= 0": (df["meta_venda"] <= 0).sum(),
    "prazo_entrega_dias < 0": (df["prazo_entrega_dias"] < 0).sum()
}

valores_invalidos = pd.DataFrame(
    valores_invalidos.items(),
    columns=["verificacao", "qtd_registros"]
)

valores_invalidos

,verificacao,qtd_registros
0,quantidade <= 0,10
1,preco_unitario <= 0,0
2,desconto < 0,0
3,desconto > 1,0
4,custo_unitario <= 0,0
5,receita < 0,0
6,lucro < 0,126
7,meta_venda <= 0,0
8,prazo_entrega_dias < 0,0


In [ ]:
# Verificar consistência da receita
# Receita esperada = quantidade * preco_unitario * (1 - desconto)

receita_calculada = df["quantidade"] * df["preco_unitario"] * (1 - df["desconto"])

# Diferença absoluta entre receita da base e receita calculada
diferenca_receita = (df["receita"] - receita_calculada).abs()

# Considerar inconsistência quando a diferença for maior que 0,01
inconsistencias_receita = df[diferenca_receita > 0.01]

print(f"Quantidade de inconsistências na receita: {len(inconsistencias_receita)}")

inconsistencias_receita[
    [
        "id_venda",
        "produto",
        "quantidade",
        "preco_unitario",
        "desconto",
        "receita"
    ]
]

In [ ]:
# Verificar consistência do lucro
# Lucro esperado = receita - (quantidade * custo_unitario)

lucro_calculado = df["receita"] - (df["quantidade"] * df["custo_unitario"])

# Diferença absoluta entre lucro da base e lucro calculado
diferenca_lucro = (df["lucro"] - lucro_calculado).abs()

# Considerar inconsistência quando a diferença for maior que 0,01
inconsistencias_lucro = df[diferenca_lucro > 0.01]

print(f"Quantidade de inconsistências no lucro: {len(inconsistencias_lucro)}")

inconsistencias_lucro[
    [
        "id_venda",
        "produto",
        "quantidade",
        "receita",
        "custo_unitario",
        "lucro"
    ]
]

In [21]:
# Verificar consistência da coluna atingiu_meta
# Se receita >= meta_venda, então deveria ser "Sim"; caso contrário, "Não"

atingiu_meta_calculado = np.where(df["receita"] >= df["meta_venda"], "Sim", "Não")

inconsistencias_meta = (df["atingiu_meta"] != atingiu_meta_calculado).sum()

print(f"Inconsistências na coluna atingiu_meta: {inconsistencias_meta}")

Inconsistências na coluna atingiu_meta: 0


In [ ]:
# Selecionar colunas numéricas relevantes para análise de outliers
colunas_numericas_analise = [
    "tempo_relacionamento_meses",
    "quantidade",
    "preco_unitario",
    "desconto",
    "receita",
    "custo_unitario",
    "lucro",
    "meta_venda",
    "prazo_entrega_dias"
]

resultado_outliers = []

for coluna in colunas_numericas_analise:
    q1 = df[coluna].quantile(0.25)
    q3 = df[coluna].quantile(0.75)
    iqr = q3 - q1
    
    limite_inferior = q1 - 1.5 * iqr
    limite_superior = q3 + 1.5 * iqr
    
    qtd_outliers = ((df[coluna] < limite_inferior) | (df[coluna] > limite_superior)).sum()
    
    resultado_outliers.append({
        "coluna": coluna,
        "q1": q1,
        "q3": q3,
        "limite_inferior": limite_inferior,
        "limite_superior": limite_superior,
        "qtd_outliers": qtd_outliers
    })

outliers_df = pd.DataFrame(resultado_outliers)

outliers_df

In [25]:
# Criar um dicionário interpretativo dos atributos da base

dicionario_dados = pd.DataFrame({
    "atributo": [
        "id_venda",
        "data",
        "mes",
        "trimestre",
        "cliente",
        "tipo_cliente",
        "tempo_relacionamento_meses",
        "faixa_renda",
        "segmento",
        "produto",
        "categoria",
        "marca",
        "linha_produto",
        "quantidade",
        "preco_unitario",
        "desconto",
        "receita",
        "custo_unitario",
        "lucro",
        "regiao",
        "cidade",
        "canal_venda",
        "vendedor",
        "meta_venda",
        "atingiu_meta",
        "prazo_entrega_dias",
        "status_entrega"
    ],
    "descricao": [
        "Identificador único da venda.",
        "Data em que a venda foi realizada.",
        "Mês numérico da venda.",
        "Trimestre em que a venda ocorreu.",
        "Identificação do cliente.",
        "Classificação do cliente como novo ou recorrente.",
        "Tempo de relacionamento do cliente com a empresa, em meses.",
        "Faixa de renda do cliente.",
        "Segmento da venda, como varejo ou atacado.",
        "Produto vendido.",
        "Categoria comercial do produto.",
        "Marca do produto.",
        "Linha comercial do produto, como econômico, padrão ou premium.",
        "Quantidade de unidades vendidas.",
        "Preço unitário do produto vendido.",
        "Percentual de desconto aplicado na venda.",
        "Valor total da venda após aplicação do desconto.",
        "Custo unitário do produto para a empresa.",
        "Lucro obtido na venda.",
        "Região onde a venda foi realizada.",
        "Cidade onde a venda foi realizada.",
        "Canal de venda utilizado, como loja ou online.",
        "Vendedor responsável pela venda.",
        "Meta de venda associada ao registro.",
        "Indica se a meta de venda foi atingida.",
        "Prazo de entrega em dias.",
        "Status da entrega, indicando se foi realizada no prazo ou com atraso."
    ]
})

dicionario_dados

In [ ]:
# Criar tabela de classificação dos tipos de dados

classificacao_tipos = pd.DataFrame({
    "atributo": [
        "id_venda",
        "data",
        "mes",
        "trimestre",
        "cliente",
        "tipo_cliente",
        "tempo_relacionamento_meses",
        "faixa_renda",
        "segmento",
        "produto",
        "categoria",
        "marca",
        "linha_produto",
        "quantidade",
        "preco_unitario",
        "desconto",
        "receita",
        "custo_unitario",
        "lucro",
        "regiao",
        "cidade",
        "canal_venda",
        "vendedor",
        "meta_venda",
        "atingiu_meta",
        "prazo_entrega_dias",
        "status_entrega"
    ],
    "tipo_pandas": [
        "int64",
        "object",
        "int64",
        "int64",
        "object",
        "object",
        "int64",
        "object",
        "object",
        "object",
        "object",
        "object",
        "object",
        "int64",
        "float64",
        "float64",
        "float64",
        "float64",
        "float64",
        "object",
        "object",
        "object",
        "object",
        "float64",
        "object",
        "int64",
        "object"
    ],
    "tipo_conceitual": [
        "Identificador",
        "Temporal",
        "Numérico discreto / temporal",
        "Numérico discreto / temporal",
        "Categórico nominal",
        "Categórico nominal binário",
        "Numérico discreto",
        "Categórico ordinal",
        "Categórico nominal",
        "Categórico nominal",
        "Categórico nominal",
        "Categórico nominal",
        "Categórico ordinal",
        "Numérico discreto",
        "Numérico contínuo",
        "Numérico contínuo percentual",
        "Numérico contínuo financeiro",
        "Numérico contínuo financeiro",
        "Numérico contínuo financeiro",
        "Categórico nominal",
        "Categórico nominal",
        "Categórico nominal",
        "Categórico nominal",
        "Numérico contínuo financeiro",
        "Categórico nominal binário",
        "Numérico discreto",
        "Categórico nominal binário"
    ]
})

classificacao_tipos

# Relatório de Diagnóstico e Compreensão de Dados

## 2. Etapa 2 - Compreensão e Diagnóstico dos Dados

A fase de diagnóstico é fundamental para garantir a integridade das análises subsequentes. A base de dados da **Comercial Insight Ltda.** compreende 1.020 registros e 27 atributos, cobrindo o ciclo operacional de vendas, clientes, logística e performance comercial no período de **01/01/2024 a 30/12/2024**.

### 2.1. Diagnóstico de Qualidade dos Dados (Data Quality)

Após uma auditoria inicial na base, foram identificados pontos críticos que exigem intervenção na etapa de pré-processamento:

* **Valores Ausentes (Missing Data):** Identificamos 51 registros (5% da base) com vacância nas colunas críticas: `faixa_renda`, `preco_unitario`, `desconto` e `custo_unitario`. A ausência ocorre de forma simultânea nessas linhas, o que aponta para uma falha sistêmica na integração de dados ou no processo de exportação do ERP/CRM.
* **Redundância e Duplicidade:** Foram detectadas 20 linhas integralmente duplicadas e 20 duplicidades específicas na chave primária `id_venda`. Considerando que a base deveria conter 1.000 transações únicas, a remoção desses registros é mandatória para evitar a inflação artificial dos indicadores financeiros.
* **Inconsistências Lógicas de Negócio:** * **Quantidade Zero:** Foram localizados 10 registros com `quantidade` igual a zero, porém apresentando `receita` e `lucro` positivos. Este é um erro grave de integridade lógica, invalidando os cálculos de preço médio e volume.
    * **Erros de Classificação Regional:** 30 registros foram classificados na região `Exterior`, apesar de as cidades listadas serem brasileiras. Isso exigirá um saneamento geográfico para não distorcer o desempenho regional.
* **Integridade Temporal e Financeira:** * As colunas de granularidade temporal (`mes` e `trimestre`) estão íntegras e coerentes com a data da venda.
    * Foram identificados 126 registros com **lucro negativo**. Diferente dos outros erros, estes casos serão mantidos para análise, pois representam a realidade operacional (vendas com prejuízo ou queima de estoque), exceto onde houver conflito com as inconsistências de quantidade zero.

### 2.2. Identificação e Descrição dos Atributos

Dada a ausência de um dicionário de dados formal, realizamos a engenharia reversa dos metadados. Os 27 atributos foram mapeados para cobrir quatro pilares principais: **Transacional** (venda, data, valores), **Entidades** (cliente, vendedor), **Logística** (entrega, prazos) e **Estratégico** (metas, atingimento).

Esta classificação é essencial para a estruturação do futuro Modelo Dimensional (Star Schema), onde separaremos o que é métrica (Fato) do que é contexto (Dimensões).

### 2.3. Classificação Técnica das Variáveis

A classificação abaixo orientará o tratamento estatístico e a escolha dos algoritmos de visualização:

| Categoria | Tipo de Dado | Atributos |
| :--- | :--- | :--- |
| **Identificadores** | Qualitativo Nominal | `id_venda`, `cliente` |
| **Temporais** | Time Series | `data`, `mes`, `trimestre` |
| **Categóricas Nominais** | Qualitativo Nominal | `tipo_cliente`, `segmento`, `produto`, `categoria`, `marca`, `regiao`, `cidade`, `canal_venda`, `vendedor`, `status_entrega` |
| **Categóricas Ordinais** | Qualitativo Ordinal | `faixa_renda`, `linha_produto`, `atingiu_meta` |
| **Numéricas Discretas** | Quantitativo Discreto | `quantidade`, `tempo_relacionamento_meses`, `prazo_entrega_dias` |
| **Numéricas Contínuas** | Quantitativo Contínuo | `preco_unitario`, `desconto`, `receita`, `custo_unitario`, `lucro`, `meta_venda` |

**Nota Técnica:** A variável `data`, embora lida inicialmente como *string* (objeto), será tipada como `datetime64` na Etapa 3 para permitir cálculos de sazonalidade e séries temporais.